File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
# 11053
# 가장 긴 증가하는 부분 수열

import sys
input = sys.stdin.readline

N = int(input())
nums = list(map(int, input().split()))
dp = [0 for _ in range(N)]

for i in range(N): # 현재 수
    for j in range(i): # 이전 수 방문
        if nums[i] > nums[j] and dp[i] < dp[j]: # 현재 값이 이전수보다 크고, dp 값이 이전 값이 더 클 경우 가져옴
            dp[i] = dp[j]
    dp[i] += 1
print(max(dp))

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N+1];
        int[] dp = new int[N+1];

        String[] inputs = br.readLine().split(" ");
        for(int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(inputs[i]);
        }

        int res = Integer.MIN_VALUE;
        for(int i = 0; i < N; i++) {
            dp[i] = 1;
            for(int j = 0; j < i; j++) {
                if (arr[j] < arr[i]) {
                    dp[i] = Math.max(dp[j]+1, dp[i]);
                }
            }
            res = Math.max(res, dp[i]);
        }

        System.out.println(res);
    }
}
